### PR TITLE
feat(client): implement batchCheck, listRelations, and non-transaction write

### PR DESCRIPTION
**Pull request by booniepepper**
## Description

* feat(client): implement batchCheck, listRelations, and non-transaction write

## References

Generated by https://github.com/openfga/sdk-generator/pull/235

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 1]
```diff
@@ -383,7 +387,29 @@ public CompletableFuture<ClientCheckResponse> check(ClientCheckRequest request,
...
      */
-    // TODO
+    public CompletableFuture<List<ClientCheckResponse>> batchCheck(
+            List<ClientCheckRequest> requests, ClientBatchCheckOptions options) {
+        int maxParallelRequests = options.getMaxParallelRequests() != null
+                ? options.getMaxParallelRequests()
+                : DEFAULT_MAX_METHOD_PARALLEL_REQS;
+        var executor = Executors.newWorkStealingPool(maxParallelRequests);
+
+        var responses = new ConcurrentLinkedQueue<ClientCheckResponse>();
```
Comment (booniepepper) [inline src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:397]: TODO: When testing look into other concurrent collection types here.

Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java]: ConcurrentLinkedQueue is the thing to do. Non-blocking beyond CAS instructions and can't find a way to break it.

File: src/test/java/dev/openfga/sdk/api/client/OpenFgaClientTest.java [hunk 2]
```diff
@@ -1086,6 +1087,50 @@ public void writeTest_deletes() throws Exception {
         mockHttpClient.verify().post(postPath).withBody(is(expectedBody)).called(1);
     }
 
+    @Test
+    public void writeTest_transactions() throws Exception {
```
[the remark this answers is not in the document]
Comment (booniepepper) [reply src/test/java/dev/openfga/sdk/api/client/OpenFgaClientTest.java:1091]: I'll add a test here

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 3]
```diff
@@ -273,19 +277,19 @@ public CompletableFuture<ClientWriteResponse> write(ClientWriteRequest request,
         configuration.assertValid();
         String storeId = configuration.getStoreIdChecked();
 
+        if (options != null && options.enableTransactions()) {
```
[the remark this answers is not in the document]
Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:280]: I'll update the names

**Comment by booniepepper** [comment]
I also need to add these to the README

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:280]: Also, I'm realizing that I also have the default backward compared to other SDKs

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 4]
```diff
@@ -273,19 +277,20 @@ public CompletableFuture<ClientWriteResponse> write(ClientWriteRequest request,
         configuration.assertValid();
         String storeId = configuration.getStoreIdChecked();
 
-        WriteRequest body = new WriteRequest();
+        if (options != null && !options.disableTransactions()) {
```
Comment (adriantam) [inline src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:280]: In other [SDKs](https://github.com/openfga/js-sdk#write-create-and-delete-relationship-tuples) and also in our README, the default behaviour if options is not provided is transaction mode.

**Review by adriantam — CHANGES_REQUESTED** [review]

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java (hunk 4 above)
Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:280]:
```suggestion
        if (options == null || !options.disableTransactions()) {
```

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 5]
```diff
@@ -277,11 +277,15 @@ public CompletableFuture<ClientWriteResponse> write(ClientWriteRequest request,
         configuration.assertValid();
         String storeId = configuration.getStoreIdChecked();
 
-        if (options != null && !options.disableTransactions()) {
-            return writeTransactions(storeId, request, options);
+        if (options == null) {
+            options = new ClientWriteOptions();
```
Comment (adriantam) [inline src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:281]: Why not just leave it as null and then treat it as writeTransactions?

Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:281]: (Which breaks a test where I had intentionally written a chunk size of 1)

Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java]: Addressing in 150b77290ea4ef975d9a5f2f98098d1236128b3d

File: src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java [hunk 6]
```diff
@@ -273,19 +277,20 @@ public CompletableFuture<ClientWriteResponse> write(ClientWriteRequest request,
         configuration.assertValid();
         String storeId = configuration.getStoreIdChecked();
 
-        WriteRequest body = new WriteRequest();
+        if (options != null && (options.disableTransactions() || options.getTransactionChunkSize() == 1)) {
```
Comment (adriantam) [inline src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java:280]: curious why we want to treat chunksize = 1 differently?

Comment (booniepepper) [reply src/main/java/dev/openfga/sdk/api/client/OpenFgaClient.java]: I got the idea from comments I read in some of the SDKs. I'm not sure if I read the comment correctly, or if behavior changes in the 1 case:

* https://github.com/openfga/dotnet-sdk/blob/be2d963baae1e38c0f7ec21323937c26b2b80954/src/OpenFga.Sdk/Client/Client.cs#L166
* https://github.com/openfga/js-sdk/blob/78eb4cfdea14455f0b63d3f5f98c78f8209a09ce/client.ts#L369

Talked offline and we will not do any special behavior when chunk size is 1 here

File: src/test/java/dev/openfga/sdk/api/client/OpenFgaClientTest.java [hunk 7]
```diff
@@ -1086,6 +1090,102 @@ public void writeTest_deletes() throws Exception {
...
```
Comment (adriantam) [inline src/test/java/dev/openfga/sdk/api/client/OpenFgaClientTest.java:1094]: Do we have any test for non transaction write?

**Review by adriantam — APPROVED** [review]

**Review by adriantam — APPROVED** [review]

**Pull request merged**